staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go

```diff
@@ -0,0 +1,550 @@
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	switch old := old.(type) {
+	case map[string]interface{}:
+		newObj, ok := new.(map[string]interface{})
+		if !ok {
+			// e.g the CRD definition has changed
+			// TODO: what's the expectation here?
+		}
+		for k, prop := range s.Properties {
+			// XImmutable and XImmutableKeys are mutually exclusive
+			// CRD validation should never let this be true
+			if prop.XImmutability != nil && prop.XKeyImmutability != nil {
+				allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "XImmutability and XKeyImmutability are mutually exclusive"))
+				continue
+			}
+
+			_, foundOld := old[k]
+			_, foundNew := newObj[k]
+
+			// XImmutability
+			// validate immutable / addOnly /removeOnly against the current property regardless of the type.
+			// XImmutability/XKeyImmutability below this level is disallowed.
+			if prop.XImmutability != nil {
+				if *prop.XImmutability == xImmutabilityImmutable {
+					if (foundOld != foundNew) || !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					}
+					continue
+				}
+				if *prop.XImmutability == xImmutabilityAddOnly {
+					if foundOld && !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is addOnly"))
+					}
+					continue
+				}
+				if *prop.XImmutability == xImmutabilityRemoveOnly {
+					if foundNew && !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is removeOnly"))
+					}
+					continue
+				}
+				continue
+			}
+
+			// Mutable array, immutable items, list type undefined / atomic
+			// The items with their respective index in the array are immutable.
+			// Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+			if prop.Type == genericTypeArray &&
+				(derefPointerToString(prop.XListType) == "" || derefPointerToString(prop.XListType) == xListTypeAtomic) &&
+				derefPointerToString(prop.Items.XImmutability) == xImmutabilityImmutable {
+
+				// TODO: check foundOld/foundNew
+				oldArr, ok := old[k].([]interface{})
+				if !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					continue
+				}
+				newArr, ok := newObj[k].([]interface{})
+				if !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					continue
+				}
+				if !isMutableArrayImmutableItems(oldArr, newArr) {
+					allErrs = append(allErrs, field.Invalid(pth.Child(s.Title), newArr, "is immutable"))
+					continue
+				}
+				continue
+			}
+
+			// Mutable array, immutable items, list type map
+			// The key-value pairs in the array are immutable, the set of keys is not.
+			// Hence, addition and removal of key-values pairs is allowed.
+			if prop.Type == genericTypeArray &&
```

Comment:
all the array logic should go to the array case of the top-level switch statement. I think you are missing non-field slices now by having it here:

```
type: array
items:
  type: array
  x-kubernetes-mutability: Immutable
```